File: cmd/remove/branch.go
package remove

import (
	"errors"
	"fmt"

	"github.com/fioncat/roxide/cmd"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/git"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/fioncat/roxide/pkg/term"
	"github.com/spf13/cobra"
)

func newBranch() *cobra.Command {
	var opts branchOptions

	c := &cobra.Command{
		Use:   "branch [NAME]",
		Short: "Remove a branch, default will remove current branch",

		Args: cobra.MaximumNArgs(1),

		ValidArgsFunction: cmd.BuildCompletion(cmd.BranchCompletion),
	}

	return cmd.Build(c, &opts)
}

type branchOptions struct {
	name string
}

func (o *branchOptions) Complete(c *cobra.Command, args []string) error {
	if len(args) > 0 {
		o.name = args[0]
	}
	return nil
}

func (o *branchOptions) Run(ctx *context.Context) error {
	repo, err := repoutils.MustGetCurrentRepo(ctx)
	if err != nil {
		return err
	}
	err = ctx.SetRepo(repo)
	if err != nil {
		return err
	}

	var isCurrent bool
	if o.name == "" {
		currentBranch, err := git.GetCurrentBranch(ctx.GetRepoPath())
		if err != nil {
			return err
		}
		o.name = currentBranch
		isCurrent = true
	} else {
		branches, err := git.ListBranches(ctx.GetRepoPath())
		if err != nil {
			return err
		}
		var found bool
		for _, branch := range branches {
			if branch.Name == o.name {
				isCurrent = branch.Current
				found = true
				break
			}
		}
		if !found {
			return fmt.Errorf("cannot find branch %q", o.name)
		}
	}

	defaultBranch, err := git.GetDefaultBranch(ctx.GetRepoPath())
	if err != nil {
		return err
	}

	if o.name == defaultBranch {
		return errors.New("cannot remove default branch")
	}

	gitCmd := git.WithPath(ctx.GetRepoPath())

	if isCurrent {
		err = gitCmd.Run("checkout", defaultBranch)
		if err != nil {
			return err
		}
	}

	err = gitCmd.Run("branch", "-D", o.name)
	if err != nil {
		return err
	}

	err = term.Confirm("Do you want to remove branch %q in remote", o.name)
	if err == nil {
		return gitCmd.Run("push", "origin", "--delete", o.name)
	}

	return nil
}

File: src/cmd/remove/branch.rs
use anyhow::{Result, bail};
use async_trait::async_trait;
use clap::Args;

use crate::cmd::Command;
use crate::cmd::complete;
use crate::config::context::ConfigContext;
use crate::debug;
use crate::exec::git::branch::{Branch, BranchStatus};
use crate::exec::git::commit::ensure_no_uncommitted_changes;

#[derive(Debug, Args)]
pub struct RemoveBranchCommand {
    pub branch: Option<String>,
}

#[async_trait]
impl Command for RemoveBranchCommand {
    async fn run(self, ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Run remove branch command: {:?}", self);

        let branches = Branch::list(ctx.git())?;
        let default_branch = Branch::default(ctx.git())?;
        let branch = match self.branch {
            Some(branch) => {
                let found = branches.iter().any(|b| b.name == branch.as_str());
                if !found {
                    bail!("cannot find branch {branch:?}");
                }
                if branch == default_branch {
                    bail!("cannot remove default branch {default_branch:?}");
                }
                branch
            }
            None => {
                let current = Branch::current(ctx.git())?;
                if current == default_branch {
                    bail!("you are in the default branch {default_branch:?}, cannot remove it");
                }
                ensure_no_uncommitted_changes(ctx.git())?;
                ctx.git().execute(
                    ["checkout", &default_branch],
                    format!("Checkout to default branch {default_branch:?}"),
                )?;
                current
            }
        };

        ctx.git().execute(
            ["branch", "-D", &branch],
            format!("Remove local branch {branch:?}"),
        )?;

        let branch = branches.into_iter().find(|b| b.name == branch);
        if let Some(branch) = branch
            && !matches!(branch.status, BranchStatus::Gone | BranchStatus::Detached)
        {
            ctx.git().execute(
                ["push", "origin", "--delete", &branch.name],
                format!("Remove remote branch {:?}", branch.name),
            )?;
        }

        Ok(())
    }

    fn complete_command() -> clap::Command {
        clap::Command::new("branch").arg(complete::branch_arg())
    }
}
